enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,298 @@
+---
+title: ovn-kubernetes-ipsec
+authors:
+  - "@mdgray"
+reviewers:
+  - TBD
+  - "@dcbw"
+approvers:
+  - TBD
+creation-date: 2020-09-08
+last-updated: 2020-09-80
+status: provisional
+see-also:
+  - "/enhancements/20190919-OVN-Kubernetes.md"
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# OVN Kubernetes IPsec Enablement
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The OVN Kubernetes network plugin uses OVN (https://www.ovn.org) to instantiate
+virtual networks for Kubernetes. These virtual networks use Geneve, a network
+virtualization overlay encapsulation protocol, to tunnel traffic across the
+underlay network between Kubernetes Nodes.
+
+IPsec is a protocol suite that enables secure network communications between
+IP endpoints by providing the following services:
+- Confidentiality
+- Authentication
+- Data Integrity
+
+OVN tunnel traffic is transported by physical routers and switches. These
+physical devices could be untrusted (devices in public network) or might be
+compromised. Enabling IPsec encryption for this tunnel traffic can prevent the
+traffic data from being monitored and manipulated.
+
+This enhancement allows cluster operators to enable traffic to be encapsulated
+by the IPsec protocol between worker nodes.
+
+## Motivation
+
+Encryption services are recommended when traffic is traversing an untrusted
+network. Encryption services may also be required for regulatory or compliance
+reasons.
+
+### Goals
+
+- Provide an option to enable IPsec encryption of all inter-worker node
+tunnel traffic across an entire Kubernetes cluster that has been configured with
+OVN Kubernetes.
+  - Assume that all nodes will be provisioned with certificates that have been
+  signed by the OpenShift Container Platform internal certificate authority
+  (CA).
+  - This configuration option will only be available on Geneve tunnels
+- All components used to implement this must be FIPs-compliant cryptographic
+components.
+- Provide an implementation that will be deployable on both Public and Private
+Cloud infrastructure. This will require that all inter-worker node traffic can
+be routed by the underlay network without reconfiguration of underlay firewall
+policies.
```

Comment:
This sentence was inspired by " If you do not use NAT, you must include encapsulation=yes in the configuration to force encapsulation. The Amazon and Azure internal cloud networks do not route IPsec ESP or AH packets. These packets must be encapsulated in UDP, and if it is configured, NAT detection configures the ESP in UDP encapsulation. If you use NAT or if you are not under the Network/Cloud-Provider limitations as described before, omit this parameter and value. " from the OpenShift admin guide. I realise now that it doesn't really reflect the underlying problem. I can reword if the comment from the OpenShift admin guide is still correct?

Reply:
(Quote is from https://docs.openshift.com/container-platform/3.11/admin_guide/ipsec.html#admin-guide-ipsec-ipsec-policy)

I really don't understand what that footnote is saying... "use NAT" for what?

> I realise now that it doesn't really reflect the underlying problem.

Yeah, other than the "use NAT" mystery, it seems like it's saying that we just need to use `encapsulation=yes` for AWS and Azure, but we don't want to use it on "real" networks. But in the 4.x solution, this should be something that we figure out automatically rather than requiring the user to choose the right option. So it's worth mentioning, but I think it's more about "we need different options for different clouds" not "... without reconfiguration of underlay firewall policies"

Reply:
"use NAT" here would be referring to NAT being between two IPsec endpoints.
But I think the guide has the cases mixed up, it should be "If you use NAT, you must include encapsulation=yes / if you do not use NAT or if you are not under.. ". ESP-in-UDP encapsulation is specifically for getting through NAT that doesn't allow IPsec headers.

Reply:
Suggest using UDP/4500 for everything to avoid even problems on bare metal.  Otherwise any device that is possibly between nodes would need to allow UDP/500 (for IKE) and proto ESP.  IMHO it's easier to document that just an UDP/4500 rule needs to be added and it will cover all cases

Reply:
Updated to say that we may need options to enable NAT traversal.